Distribution specific override functionality based on class inheritance

joohoi opened the pull request:
Class inheritance based approach to distro specific overrides. 

~The route I took involves some constructor magic, and in-function includes in constants to avoid cyclic imports which would have been impossible to avoid without it. Override classes inherit `configurator.OverrideConfigurator` which has the constructor magic in place in conjunction with `ApacheConfigurator`. This is to avoid unnecessary boilerplate in the actual override classes.~

**How it works:**
The certbot-apache plugin entrypoint has been changed to `entrypoint.ENTRYPOINT` which is a variable containing appropriate override class for system, if available.

Override classes register themselves using decorator `override.register()` which takes a list of distribution fingerprints (ID & LIKE variables in /etc/os-release, or `platform.linux_distribution()` as a fallback). These end up as keys in `dict` `override.OVERRIDE_CLASSES` and values for the keys are references to the class that called the decorator, hence allowing self-registration of override classes when they are imported. The only file importing these override classes is `entrypoint.py`, so adding new override classes would need only one `import` in addition to the actual override class file.


~When an instance of `ApacheConfigurator` is created, the constructor method calls `constants.get_configurator()` which checks distribution fingerprint, and creates a new override class if one is available for distribution in question, and returns it for the `ApacheConfigurator` which then returns this override class (that inherits `ApacheConfigurator` ) instead of a instance of itself.~

**Generic changes:**
 - Parser initialization has been moved to separate class method, allowing easy override where needed.
 - Cleaned up `configurator.py` a bit, and moved some helper functions to newly created `apache_util.py`
 - Split Debian specific code from `configurator.py` to `debian_override.py`
 - Changed `define_cmd` to `apache_cmd` because the parameters are for every distribution supporting this behavior, and we're able to use the value to build the additional configuration dump commands.
 - Moved `add_parser_mod()` from configurator to parser `add_mod()`
 - Added two new configuration dump parsing methods to `update_runtime_variables()` in parser: `update_includes()` and `update_modules()`.
 - Changed `init_modules()` in parser to accommodate the changes above. (ie. don't throw existing self.modules out).
 - Moved OS based constants to their respective override classes.
 - Refactored configurator class discovery in tests to help easier test case creation using distribution based override configurator class.
 - `tests.util.get_apache_configurator()` now takes keyword argument `os_info` which is string of the desired mock OS fingerprint response that's used for picking the right override class.

This PR includes two major generic additions that should vastly improve our parsing accuracy and quality:

1) Includes are parsed from config dump from httpd binary. This is mandatory for some distributions (Like OpenSUSE) to get visibility over the whole configuration tree because of Include statements passed on in command line, and not via root httpd.conf file.
2) Modules are parsed from config dump from httpd binary. This lets us jump into correct IfModule directives if for some reason we have missed the module availability (because of one being included on command line or such).


**Distribution specific changes**
Because of the generic changes, there are two distributions (or distribution families) that do not provide such functionality, so it had to be overridden in their respective override files. These distributions are:

1) CentOS, because it deliberately limits httpd binary stdout using SELinux as a feature. We are doing opportunistic config dumps here however, in case SELinux enforcing is off.
2) Gentoo, because it does not provide a way to invoke httpd with command line parsed from its specific configuration file. Gentoo relies heavily on Define statements that are passed over from APACHE2_OPTS variable /etc/conf.d/apache2 file and most of the configuration in root Apache configuration are dependent on these values.

**Debian**
 - Moved the Debian specific parts from configurator.py to Debian specific override.

**CentOS**
 - Parsing of `/etc/sysconfig/httpd` file for additional Define statements. This could hold other parameters too, but parsing everything off it would require a full Apache lexer. For CLI parameters, I think Defines are the most common ones. This is done in addition of opportunistic parsing of httpd binary config dump.
 - Added CentOS default Apache configuration tree for realistic test cases.

**Gentoo**
 - Parsing Defines from `/etc/conf.d/apache2` variable `APACHE2_OPTS`, which holds additional Define statements to enable certain functionalities, enabling parts of the configuration in the Apache2 DOM. This is done instead of trying to parse httpd binary configuration dumps.
 - Added default Apache configuration from Gentoo to testdata, including `/etc/conf.d/apache2` file for realistic test cases.

bmw commented on certbot-apache/certbot_apache/tests/centos_test.py:
```diff
@@ -0,0 +1,138 @@
+        self.assertTrue("mock_define" in self.config.parser.variables.keys())
+        self.assertTrue("mock_define_too" in self.config.parser.variables.keys())
+        self.assertTrue("mock_value" in self.config.parser.variables.keys())
+        self.assertEqual("TRUE", self.config.parser.variables["mock_value"])
```
You should add:
```
if __name__ == "__main__":
    unittest.Main()
```
here.

bmw commented on certbot-apache/certbot_apache/tests/generic_test.py:
```diff
@@ -0,0 +1,106 @@
+    def test_enable_mod_unsupported(self):
+        self.assertRaises(errors.MisconfigurationError,
+                          self.config.enable_mod,
+                          "whatever")
```
You should add:
```
if __name__ == "__main__":
    unittest.Main()
```
here.

bmw commented on certbot-apache/certbot_apache/tests/gentoo_test.py:
```diff
@@ -0,0 +1,103 @@
+        self.config.parser.variables = {}
+        self.config.parser.update_runtime_variables()
+        for define in defines:
+            self.assertTrue(define in self.config.parser.variables.keys())
```
You should add:
```
if __name__ == "__main__":
    unittest.Main()
```
here.

bmw commented on certbot-apache/certbot_apache/tests/generic_test.py:
```diff
@@ -0,0 +1,106 @@
+import shutil
+
+import mock
+import tempfile
```
nit: tempfile is part of the standard library and should be moved with the first group of imports.

bmw commented:
Some test farm tests are currently broken due to the same issues as #5205. I want to do some more digging into why this is suddenly a problem, but I'll fix them soon and report back.

bmw commented:
Running test farm tests by merging this PR and #5216 into a single branch, tests fail on Fedora 23. Here's the [log](https://gist.github.com/5c75e51921e683eb7f55c0536c303e06).

joohoi commented on .pylintrc:
```diff
@@ -41,7 +41,7 @@ load-plugins=linter_plugin
 # no Warning level messages displayed, use"--disable=all --enable=classes
 # --disable=W"
-disable=fixme,locally-disabled,abstract-class-not-used,abstract-class-little-used,bad-continuation,too-few-public-methods,no-self-use,invalid-name,too-many-instance-attributes,cyclic-import
+disable=fixme,locally-disabled,abstract-class-not-used,abstract-class-little-used,bad-continuation,too-few-public-methods,no-self-use,invalid-name,too-many-instance-attributes,cyclic-import,duplicate-code
```
The reason was setUp methods in tests, mocking the environment through distribution fingerprinting and mock constants. I moved the mocks to `get_apache_configurator` in tests.util.
This also helped us get rid of a lot of test setup boilerplate! 

Other part was doing the the same testcase in all of the override tests where we check that all the defined vhosts are found in the configuration. I just renamed variables in these tests, as I think it's better to keep the tests in the actual test files instead of running test from external file.

joohoi commented on certbot-apache/certbot_apache/apache_util.py:
```diff
@@ -0,0 +1,96 @@
+    """
+    return_vars = {}
+    # Get list of words in the variable
+    a_opts = util.get_var_from_file(varname, filepath).split(" ")
```
It does, this is fixed

joohoi commented on certbot-apache/certbot_apache/configurator.py:
```diff
@@ -1943,79 +1890,7 @@ def cleanup(self, achalls):
-        internal_path.append(internal_path_part)
-
-    return file_path, "/".join(reversed(internal_path))
+            self.parser.reset_modules()
```
It is actually done here: https://github.com/certbot/certbot/blob/373914fe719c50ca8474ca867e059b57eaeb6b80/certbot-apache/certbot_apache/configurator.py#L1891

But we just want to reset the parser state, ie: reading the active modules again from the configuration (and optionally from binary config dump).

bmw replied:
Sorry my comment wasn't very clear. I was referring to the `IInstaller` methods that revert changes made to the Apache config using the reverter class. Since we're rolling back the config to a previous state, the loaded modules may have changed. This is `rollback_checkpoints` and `recovery_routine` methods.

joohoi replied:
Ah, yeah. Ultimately we should,  that should most likely be done at the same time when/if we restructure the ApacheConfigurator <-> AugeasConfigurator <-> ApacheParser relations.

This is currently the only runtime rollback we're doing, so it's not very urgent.

bmw replied:
I agree it's not necessary with the way Certbot currently uses its plugins, but this could cause bugs if this changes in the future. We should make an issue so we don't forget about this.

joohoi replied:
I created issue #5227 to track this.

joohoi commented on certbot-apache/certbot_apache/override_centos.py:
```diff
@@ -0,0 +1,34 @@
+
+    def update_runtime_variables(self, *args, **kwargs):
+        """ Override for update_runtime_variables for custom parsing """
+        # Opportunistic, works if SELinux not enforced
```
Correct, the tests I ran in my virtual CentOS boxes resulted exactly in that.

joohoi commented on certbot-apache/certbot_apache/override_gentoo.py:
```diff
@@ -0,0 +1,36 @@
+
+    def update_runtime_variables(self):
+        """ Override for update_runtime_variables for custom parsing """
+        self.parse_sysconfig_var()
```
Yeah, Gentoo doesn't provide a way to pass parameters to httpd process with command line composed out of it's own specific configuration style.

joohoi commented on certbot-apache/certbot_apache/parser.py:
```diff
@@ -103,50 +100,62 @@ def add_include(self, main_config, inc_path):
-        """"
+        """Update Includes, Defines and Includes from httpd config dump data"""
+        self.update_defines()
+        self.update_includes()
```
These are httpd binary config dump calls, so they won't be affected by actual DOM parsing. The method parsing the actual configuration gets called later on.

bmw replied:
Oh really? I thought that was done when `parse_file` was called on the file. Which method does the actual config parsing?

joohoi replied:
The actual place where we parse the DOM is `parse_modules()`, previously named as `init_modules()`. And you are right, I never stopped to think about it, but we really aren't updating modules when including new files. I think calling the binary config dumps at this point wouldn't be necessary, but we definately should call `parse_modules()` to catch new LoadModule statements!

bmw replied:
I see. Apparently I don't fully understand how this works. Where do we try to determine if directives in `<IfModule>` blocks should be included? Do we handle this or does Augeas do this for us? If the latter, where do we tell Augeas about the included modules?

joohoi replied:
That's done in `_exclude_dirs()` which is called from `find_dir()`. It's handled manually.

joohoi commented on certbot-apache/certbot_apache/tests/centos_test.py:
```diff
@@ -0,0 +1,138 @@
+# pylint: disable=duplicate-code
```
We can, and the line wasn't working because of pylint bug that didn't allow disabling of duplicate-code locally, only throgh global configuration, but I've worked around the need for that already.

joohoi commented on certbot-apache/certbot_apache/tests/centos_test.py:
```diff
@@ -0,0 +1,138 @@
+            with mock.patch(
+                "certbot_apache.constants.os_constant") as mock_c:
+                mock_c.side_effect = mock_os_constant
+                self.config = util.get_apache_configurator(
```
I did this in order to remove the duplicate-code pylint disabling

joohoi commented on certbot-apache/certbot_apache/tests/configurator_test.py:
```diff
@@ -399,9 +390,12 @@ def test_deploy_cert_enable_new_vhost(self):
-        self.config = util.get_apache_configurator(
-            self.config_path, self.vhost_path, self.config_dir,
-            self.work_dir, version=(2, 4, 16))
+        with mock.patch("certbot.util.get_os_info") as mock_osi:
```
This is in fact a problem with current design, where the OS gets fingerprinted every time `constants.os_constant("whatever")` is called to get a distro based variable. I really like the idea of moving all of it to configurator (overrides), and this would make these test hacks unnecessary.

joohoi commented on certbot-apache/certbot_apache/override_centos.py:
```diff
@@ -0,0 +1,34 @@
+    """CentOS specific ApacheParser override class"""
+    def __init__(self, *args, **kwargs):
+        self.sysconfig_filep = "/etc/sysconfig/httpd"
+        super(CentOSParser, self).__init__(*args, **kwargs)
```
For some reason, switching them around, will cause `sysconfig_filep` variable being unavailable in the class. This was new to me, and I didn't dig around for the root reason, but instead decided to leave them as they are currently.

joohoi commented:
I will look into the Fedora 23 problem more closely, even though it's EOLed almost a year ago. I want to know what's going on. Is the serverfarm test suite keeping up with the rapid release cycle (ie. are we running against the most recent ones? There's 27 in beta, so new one is upon us soon enough too).

joohoi commented on certbot-apache/certbot_apache/configurator.py:
```diff
@@ -138,6 +139,17 @@ def __init__(self, *args, **kwargs):
                               "ensure-http-header": self._set_http_header,
                               "staple-ocsp": self._enable_ocsp_stapling}
 
+    def __new__(cls, *args, **kwargs):
```
I really like the idea! The only part of the current implementation that I dislike is the `__new__` hacks. What would you suggest for a good entrypoint if we take this route?

bmw replied:
That's a good idea. Initially I was thinking it'd be a standalone function in `configurator.py` or a static method on `ApacheConfigurator`, but keeping all of this functionality to the metaclass is a nice idea.

joohoi commented on certbot-apache/certbot_apache/constants.py:
```diff
@@ -2,13 +2,59 @@
+    os_info = util.get_os_info()
+    override_class = None
+    try:
+        override_class = OVERRIDES[os_info[0].lower()]
```
I like this too! And it would be a good fit with factory implementation as well. I'll start the work toward this goal!

bmw commented:
Unfortunately our test farm tests targets are outdated and Fedora 23 is the only Fedora distro we're currently testing against. I'll get this updated.

joohoi commented on certbot-apache/certbot_apache/override_debian.py:
```diff
@@ -0,0 +1,121 @@
+""" Distribution specific override class for Debian family (Ubuntu/Debian) """
+import logging
+import os
```
This is fixed

joohoi commented:
Ok, so I implemented the override class discovery in the way we discussed over IRC. At this point the PR also moves all the OS constants to their respective override classes. Unfortunately, duplicate-code is disablet in pylintrc again, because it triggers on the OS_DEFAULTS dictionaries in the different override classes. We can re-enable it, but we would have to override only specific constant vars in the override classes if we do so. I agree that it is duplicate code, but it affects the view over distro based constants quite a bit.

There's few things to fix in the apache-compatibility tests, but the PR is ready for next round of review.

joohoi commented:
The compatibility tests are now fixed, and I updated the description of this PR, but left the deprecated parts of description in place with ~strikethrough~ for reflection.

bmw commented on certbot-compatibility-test/certbot_compatibility_test/configurators/apache/common.py:
```diff
@@ -56,13 +55,14 @@ def load_config(self):
         """Prepares the Apache plugin for testing"""
-        for k in constants.CLI_DEFAULTS_DEBIAN.keys():
-            setattr(self.le_config, "apache_" + k, constants.os_constant(k))
+        for k in override_debian.DebianConfigurator.OS_DEFAULTS.keys():
```
I actually think we should use `entrypoint.ENTRYPOINT` here to use the correct configurator for the environment the tests are run in. I made this change locally and tests still passed.

joohoi replied:
You are right, I'll make the change. All the tests are pinned to a distribution, and no fingerprinting happens runtime anymore so this isn't a big issue *currently* , but might be one in the future.

bmw replied:
Yeah. I actually didn't realize the code had changed as much as it had, but when I initially wrote the compatibility test, the plan was we'd run it on different distros with different setups which would work by providing the tests metadata about the configs.

Unfortunately, we're not currently making use of that, but maybe we should someday!

joohoi commented on certbot-apache/certbot_apache/configurator.py:
```diff
@@ -85,27 +87,53 @@ class ApacheConfigurator(augeas_configurator.AugeasConfigurator):
 
     description = "Apache Web Server plugin - Beta"
 
+    OS_DEFAULTS = dict(
```
I do agree, and looked into it a bit when implementing this, but for the sake of keeping this already pretty large PR from growing I decided to leave it for later.

bmw replied:
> I do agree, and looked into it a bit when implementing this, but for the sake of keeping this already pretty large PR from growing I decided to leave it for later.

Great. I created #5289.

> This would and should go to the same issue and development we discussed in IRC: letting users to pick override classes from CLI if fingerprinting fails. What I also want to accomplish with this would be support for multiple Apache installations in a system. Ie. giving better UX to choose root filepath, apachectl path and so on. Because of the current list() nature of the cmds it's not trivial to provide a CLI flag.

I agree we should do that work as well. As long as the new system is designed carefully with that kind of functionality in mind, they wouldn't have to be done together to keep PRs small, but it probably doesn't matter much as the changes shouldn't be that large.

I completely agree the Apache command line flags need to be reworked. Just to state this explicitly, we should try to preserve backwards compatibility and deprecate things slowly if we make any breaking changes.

joohoi commented on certbot-apache/certbot_apache/override_arch.py:
```diff
@@ -0,0 +1,34 @@
+    """Arch Linux specific ApacheConfigurator override class"""
+
+    OS_DEFAULTS = dict(
+        server_root="/etc/apache2",
```
Oh wow, nice catch, and an embarrassing mistake from my part!

bmw replied:
No worries. In a PR with 5000+ lines of churn, a few mistakes will be made. That's what reviews are for!

joohoi commented on certbot-apache/certbot_apache/tests/configurator_test.py:
```diff
@@ -35,20 +35,9 @@ class MultipleVhostsTest(util.ApacheTest):
-            self.config = util.get_apache_configurator(
-                self.config_path, None, self.config_dir, self.work_dir)
-            self.config = self.mock_deploy_cert(self.config)
+        self.config = util.get_apache_configurator(
```
I looked quickly into it previously, and it seemed like a lot of work, as a large chunk of the "old tests" make assumptions based on the distro. Now when taking a closer look I was able to separate the Debian tests with minimal duplication, so this is now taken care of.

bmw replied:
Awesome. Thanks for doing this.

I can imagine it might have been a little tedious but I think it helps clean things up and make the purpose of different files more clear.

joohoi commented on certbot-apache/certbot_apache/override.py:
```diff
@@ -0,0 +1,15 @@
+""" Registry and decorator for override classes """
+OVERRIDE_CLASSES = {}
+
+def register(distros):
```
No problem at all! I was going to argue about this on the basis of having a clear and easy way to add new os fingerprints in the override class itself, so it'd be obvious from the override class viewpoint. But thinking about it, it's much harder to actually figure out which override class would be responsible for `insert_your_os_flavor_here` without a concise list of fingerprints. That said, I think this is the best way to do it.

joohoi commented:
All the requested changes have been made, and the problem with Fedora 23 was happening because update_includes() didn't prepopulate the includes list with the `Include` and `IncludeOptional` statements found from the actual configuration prior to doing the `apachectl -t -D DUMP_INCLUDES`.

The binary configuration dump would then add specific files found to the Augeas tree (eg: `/path/to/file1.conf` and `/path/to/file2.conf`) and the actual iteration later would then add `/path/to/*.conf`. Some versions of Augeas, including the one shipped with Fedora 23 don't play well when loading a single file via multiple statements (*.conf AND file1.conf for example).

This was simply fixed by populating the parser paths first from the actual configuration file, which is the source of wildcard includes, while binary configuration dump gives fully resolved paths. The later additions check if path is already parsed, and doesn't add new ones to the list if so.

bmw commented on certbot-apache/certbot_apache/parser.py:
```diff
@@ -163,15 +169,54 @@ def update_runtime_variables(self):
+        # Find_dir iterates over configuration for Include and IncludeOptional
+        # directives to make sure we see the full include tree present in the
+        # configuration files
+        _ = self.find_dir("Include")
```
Two questions here:
1. Do we also need to look for `IncludeOptional`? If not, why not?
2. Is it possible for the reverse problem you described to happen? That is, if the user's config includes `/path/to/file.conf`, will `DUMP_INCLUDES` ever include a wildcard like `/path/to/*.conf` in its output? This has the potential to break things for users where the Apache plugin currently works so I want to make sure we're not expecting breakage here.

Also, a nit. Personally, with the return value not being used I'd just change this line to `self.find_dir("Include")` and let Python automatically ignore the return value. If you like this better for some reason, feel free to leave it though.

joohoi replied:
1) Yeah, actually `find_dir()` searches for both `Include` and `IncludeOptional` statements *in addition* to the provided search directive. So it would have worked perfectly with whatever argument, but I decided to use `"Include"` for readability.
2) No, `DUMP_INCLUDES` always returns the actual files that have been loaded.

Actually I'd prefer to leave it as it is for readability. I usually do the very thing you described, but decided to leave it in this case because the method is designed to actually return something, and it's used here in a bit hackish way against its intended purpose. We should refactor that functionality out of it at some point in the future.

bmw commented on certbot-apache/certbot_apache/tests/gentoo_test.py:
```diff
@@ -0,0 +1,86 @@
+import os
+import unittest
+
+from certbot_apache import override_gentoo  # pylint: disable=unused-import
```
You can remove the `pylint` comment here as we use this module on line 57.

joohoi commented on certbot-apache/certbot_apache/override_arch.py:
```diff
@@ -0,0 +1,31 @@
+        le_vhost_ext="-le-ssl.conf",
+        handle_mods=False,
+        handle_sites=False,
+        challenge_location="/etc/apache2/vhosts.d",
```
Of course there had to be something still left to clean :)

joohoi commented:
Yeah, the few small fixes should now be in. I'm pretty confident with the Fedora 23 fix as it pretty much boils down to in which order we add the includes in. It goes from the broader scope down to specific ones.

bmw commented:
Tests are going to fail until #5291 is merged. I'm cancelling the Travis build but will kick it when they're expected to pass again.
